platform/bukkit/plugin/src/main/java/me/lokka30/treasury/plugin/bukkit/hooks/EconomyHook.java

```diff
@@ -0,0 +1,331 @@
+import org.bukkit.OfflinePlayer;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class EconomyHook implements TreasuryPAPIHook {
+
+    private EconomyProvider provider;
+    private final DecimalFormat format = new DecimalFormat("#,###");
+    private final PAPIExpansion expansion;
+    private final TreasuryBukkit plugin;
+    private final String k, m, b, t, q;
+    private Baltop baltop;
+
+    public EconomyHook(PAPIExpansion expansion, TreasuryBukkit plugin) {
+        this.expansion = expansion;
+        this.plugin = plugin;
+        this.k = expansion.getString("formatting.thousands", "k");
+        this.m = expansion.getString("formatting.millions", "m");
+        this.b = expansion.getString("formatting.billions", "b");
+        this.t = expansion.getString("formatting.trillions", "t");
+        this.q = expansion.getString("formatting.quadrillions", "q");
+    }
+
+    @Override
+    public boolean setup() {
+        Optional<Service<EconomyProvider>> serviceOpt = ServiceRegistry.INSTANCE.serviceFor(
+                EconomyProvider.class);
+        if (serviceOpt.isPresent()) {
+            provider = serviceOpt.get().get();
+            this.baltop = new Baltop(
+                    expansion.getBoolean("baltop.enabled", false),
+                    expansion.getInt("baltop.cache_size", 100),
+                    expansion.getInt("baltop.cache_delay", 60),
+                    provider
+            );
+            this.baltop.start(plugin);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean canRegister() {
+        return provider != null;
+    }
+
+    @Override
+    public @Nullable String onRequest(
+            @Nullable OfflinePlayer player, @NotNull String param
+    ) {
+        if (provider != null) {
+            if (!baltop.isEnabled() && param.startsWith("top_")) {
+                return (param.startsWith("top_balance")) ? "0" : "";
+            }
+
+            if (param.contains("top_balance")) {
+                if (param.startsWith("top_balance_fixed")) {
+                    String currencyId = param.replace("top_balance_fixed_", "");
+                    if (currencyId.isEmpty()) {
+                        currencyId = provider.getPrimaryCurrencyId();
+                    }
+                    BigDecimal topBalanceFixed = baltop.getTopBalance(currencyId);
+                    if (topBalanceFixed == null) {
+                        return "0";
+                    } else {
+                        return String.valueOf(topBalanceFixed.longValue());
+                    }
+                }
+                if (param.startsWith("top_balance_formatted")) {
+                    Locale locale = Locale.ENGLISH;
+                    if (player != null && player.isOnline()) {
+                        locale = Locale.forLanguageTag(player.getPlayer().getLocale());
```

Comment:
Note: afaik the client's language is not ever a real language tag - it uses `en_GB` etc. instead of `en-GB`, and allows pretty much anything via custom resource pack. The default client also includes several joke languages such as Pirate (`en_PT`) that may also cause trouble.

Reply:
I elected to bandaid this by replacing `_` with `-` but we may wish to split it out to a more robust utility method.